platform/security/idp/security-idp-server/src/main/java/org/codice/ddf/security/idp/server/IdpEndpoint.java

```diff
@@ -232,6 +271,133 @@ public void init() {
         new ResponseUnmarshaller());
   }
 
+  @Override
+  public Set<String> getActiveSessions() {
+    return cookieCache.getAllSamlSubjects(securityManager);
+  }
+
+  @Override
+  public void invalidateSession(String subjectName) {
+    String cacheKey = cookieCache.getCacheKeyBySubjectName(subjectName, securityManager);
+    if (cacheKey == null) {
+      LOGGER.debug(
+          "No cache element found for subject name {}; skipping invalidation.", subjectName);
+      return;
+    }
+
+    LogoutState logoutState = new LogoutState(getActiveSps(cacheKey));
+
+    // TODO: 1/3/18 sufficient?
+    logoutState.setOriginalIssuer("");
+
+    logoutState.setNameId(subjectName);
+    logoutState.setOriginalRequestId(UUID.randomUUID().toString());
+
+    String relayState = UUID.randomUUID().toString();
+    logoutState.setInitialRelayState(relayState);
+    logoutStates.encode(cacheKey, logoutState);
+
+    cookieCache.removeSamlAssertion(cacheKey);
+
+    asyncLogoutService.submit(() -> asyncBackchannelLogout(cacheKey, logoutState));
+  }
+
+  private void asyncBackchannelLogout(String cacheKey, LogoutState logoutState) {
+    ExecutorService executorService =
+        Executors.newFixedThreadPool(
+            4, StandardThreadFactoryBuilder.newThreadFactory("sessionLogout"));
+    Map<String, Future<?>> futures = new HashMap<>();
+    for (Optional<String> nextTarget = logoutState.getNextTarget();
+        nextTarget.isPresent();
+        nextTarget = logoutState.getNextTarget()) {
```

Comment:
❓ Not sure if I missed it, but do we already handle removing the client that initiated the request? (in the IdP client code we invalidate the session before sending this off, so trying to log out of them again via this logic would be unnecessary)

Reply:
This method isn't invoked for client-initiated logouts. It is only invoked for IdP-initiated logouts and sends the SOAP logout request to all attached SPs.

Reply:
How long have we supported IDP initiated logouts? Always?